reader_concurrency_semaphore.cc

```diff
@@ -240,6 +255,22 @@ class reader_permit::impl
         _semaphore.signal(res);
     }
 
+    future<resource_units> request_memory(size_t memory) {
+        auto res = _semaphore.request_memory(*this, memory);
+        if (res.available()) {
+            res.get();
+            assert(!_requested_memory);
+            assert(_state != state::waiting_for_memory);
+            return make_ready_future<resource_units>(resource_units(reader_permit(shared_from_this()), {0, ssize_t(memory)}));
+        } else {
+            _requested_memory += memory;
+            return res.then([this, memory] {
+                assert(!_requested_memory);
+                return resource_units(reader_permit(shared_from_this()), {0, ssize_t(memory)}, resource_units::already_consumed_tag{});
+            });
+        }
```

Comment:
Note sure why the two paths (checking future::available() should be left to low-level code). 

I also don't like the assert. Can't someone sneak in between res becoming available and the continuation executing?

Reply:
Hmm, turns out I needed these games with `_requested_memory` and checking the future only because I was too lazy to split `on_admission()` so there is a separate one for the admission and memory queues.
Will fix in v6.